include/ignition/math/Polynomial3.hh

```diff
@@ -0,0 +1,244 @@
+      /// \brief Constructor
+      public: Polynomial3() = default;
+
+      /// \brief Constructor
+      /// \param[in] _coeffs coefficients c0 through c3, left to right
+      public: explicit Polynomial3(Vector4<T> _coeffs)
+      : coeffs(std::move(_coeffs))
+      {
+      }
+
+      /// \brief Make a constant polynomial
+      /// \return a p(x) = `_value` polynomial
+      public: static Polynomial3 Constant(T _value)
+      {
+        return Polynomial3(Vector4<T>(0., 0., 0., _value));
+      }
+
+      /// \brief Get the polynomial coefficients
+      /// \return this polynomial coefficients
+      public: const Vector4<T> &Coeffs() const { return this->coeffs; }
+
+      /// \brief Evaluate the polynomial at `_x`
+      /// For non-finite `_x`, this function
+      /// computes p(z) as z tends to `_x`.
+      /// \param[in] _x polynomial argument
+      /// \return the result of evaluating p(`_x`)
+      public: T Evaluate(const T &_x) const
+      {
+        using std::isfinite, std::abs, std::copysign;
+        if (!isfinite(_x))
+        {
+          const T epsilon =
+              std::numeric_limits<T>::epsilon();
+          if (abs(this->coeffs[0]) >= epsilon)
+          {
+            return _x * copysign(T(1.), this->coeffs[0]);
+          }
+          if (abs(this->coeffs[1]) >= epsilon)
+          {
+            return copysign(_x, this->coeffs[1]);
+          }
+          if (abs(this->coeffs[2]) >= epsilon)
+          {
+            return _x * copysign(T(1.), this->coeffs[2]);
+          }
+          return this->coeffs[3];
+        }
+        const T _x2 = _x * _x;
+        const T _x3 = _x2 * _x;
+
+        return (this->coeffs[0] * _x3 + this->coeffs[1] * _x2 +
+                this->coeffs[2] * _x + this->coeffs[3]);
+      }
+
+      /// \brief Call operator overload
+      /// \see Polynomial3::Evaluate()
+      public: T operator()(const T &_x) const
+      {
+        return this->Evaluate(_x);
+      }
+
+      /// \brief Compute polynomial minimum in an `_interval`
+      /// \param[in] _interval polynomial argument interval to check
+      /// \return the polynomial minimum in the given interval,
+      ///   or NaN if the interval is empty
+      public: T Minimum(const Interval<T> &_interval) const
+      {
+        if (_interval.Empty())
+        {
+          return std::numeric_limits<T>::quiet_NaN();
+        }
+        using std::min, std::abs, std::sqrt;  // enable ADL
+        constexpr T epsilon = std::numeric_limits<T>::epsilon();
+        // For open intervals, assume continuity in the limit
+        T minimum = min(this->Evaluate(_interval.LeftValue()),
+                        this->Evaluate(_interval.RightValue()));
+        if (abs(this->coeffs[0]) >= epsilon)
+        {
+          // cubic poylnomial
+          const T a = this->coeffs[0] * T(3.);
+          const T b = this->coeffs[1] * T(2.);
+          const T c = this->coeffs[2];
+
+          const T discriminant = b * b - T(4.) * a * c;
```

Comment:
I haven't looked closely at the cubic polynomial calculations in this block yet, but this stood out to me since it's the quadratic discriminant, not the [cubic one](https://en.wikipedia.org/wiki/Cubic_equation#Discriminant). This could be related to the error in the tests that I pointed out below.

Reply:
Note we're not trying to compute the roots of the polynomial, but the roots of its first derivative in order to find all local minima. I've added some comments in https://github.com/ignitionrobotics/ign-math/pull/393/commits/0108ccfa1bcfe77c92ea6d1c68d30e61d554b491 to make it clearer.